rules/external_binary/def.bzl

```diff
@@ -12,22 +12,43 @@ def _external_binary_impl(ctx):
     else:
         fail("Unsupported operating system: {}".format(ctx.os.name))
 
+    url = info.get("url")
     args = {
-        "url": info.get("url"),
+        "url": url,
         "sha256": info.get("sha256", ""),
         "output": "{name}/{name}".format(name = ctx.attr.name),
     }
-    if info.get("is_compressed", False):
+    if url.endswith(".zip") or url.endswith(".tar.gz") or url.endswith(".tgz"):
         ctx.download_and_extract(**args)
+        flatten = [
+            "sh", "-c",
+            """
+            set -o errexit -o pipefail
+            mkdir '{name}_tmp'
+            find '{name}' -mindepth 2 -type f -print0 \
+                | xargs -0 -L50 mv --target-directory '{name}_tmp'
+            rm -rf '{name}'
+            mv '{name}_tmp' '{name}'
+            """.format(
+                name = ctx.attr.name,
+            )
+        ]
+        res = ctx.execute(flatten)
+        if res.stdout != "":
+            print("STDOUT:")
+            print(res.stdout)
+        if res.stderr != "":
+            print("STDERR:")
+            print(res.stderr)
```

Comment:
This _seems_ to work, but I'm not sure this covers all the cases (and I think it needs some style fixups):

```python
    url = info.get("url")
    args = {
        "url": url,
        "sha256": info.get("sha256", ""),
    }
    if any([url.endswith(suffix) for suffix in [".zip", ".tar.gz", ".tgz"]]):
        ctx.download_and_extract(output="{name}/{name}_out".format(name = ctx.attr.name), **args)
        build_contents = """
        package(default_visibility = ["//visibility:public"])
        filegroup(name = "{name}", srcs = glob(["**/{name}*"]))
        """.format(name = ctx.attr.name)
    else:
        args["executable"] = True
        ctx.download(output="{name}/{name}".format(name = ctx.attr.name), **args)
        build_contents = """
        package(default_visibility = ["//visibility:public"])
        exports_files(glob(["**/*"]))
        """.format(name = ctx.attr.name)

    build_contents = '\n'.join([x.lstrip(' ') for x in build_contents.splitlines()])
    ctx.file("{name}/BUILD.bazel".format(name = ctx.attr.name), build_contents)
```

Reply:
This didn't work as-is but gave me some light on how to do it. A `filegroup` cannot be executable, but we can have a `genrule` that is executable that takes the `filegroup` as input.

I added a `cmd_bash` and a `cmd_bat` that should cover all platforms that we want to support.